### PR TITLE
change(web): modularization, refinement of setActiveKeyboard 🧩

### PR DESCRIPTION
**Pull request by jahorton**
This PR focuses on the `setActiveKeyboard` API method and the methods that it triggers.  These previously relied on explicit references to many of the higher-level, semi-global engine components and obtained their root object via the namespaced `com.keyman.singleton` global, so some cleanup here was quite necessary.

Also of note is that the method was written in a very "imperative" manner.   Depending upon the state of the engine, there are four different conditions and control paths used to load keyboards:

1.  "Load the 'default' keyboard"
    - which differs between the 'desktop' and 'phone', 'tablet' form factors.
2.  "Load the current keyboard, but for a different language" (or even the same one, actually; it's treated the same)
3.  "Load that keyboard we loaded previously"
4.  "Load a keyboard we haven't loaded yet"

Further complicating _that_:  `kmw.beforekeyboardchange` is documented to allow cancellation of a keyboard-change operation; of note is that it is raised _twice_ for condition 4 above - once before fetching the keyboard script, once after (but before actual activation).

Finally, standard website-hosted uses of the browser expect to track the newly-set keyboard in a persistent cookie for future visits... and it shouldn't matter which of the four paths is taken to load it.

With all those moving parts, I opted to break the main loading method down into a few pieces to help reduce the complexity:
- `ContextManagerBase.prepareKeyboardForActivation` determines which of the four control paths to follow and _prepares_ the keyboard accordingly.  It is made available via Promise due to condition 4.
- `ContextManagerBase.activateKeyboard` calls that method, and _then_ handles the `kmw.beforekeyboardchange` wrinkle in a centralized manner.  It also does the actual keyboard change and associated `kmw.keyboardchange` events.
- The variant of `activateKeyboard` for the `app/browser` target additionally saves the keyboard cookie once the keyboard-change is confirmed, again in a centralized manner.

----

In the process of this work, I noticed a few interesting edge cases that needed some attention.

#### Related issue

#8482 was discovered at this time.  Since I had to handle the existing event calls and maintain their timing within the keyboard-activation process, which was the trickiest aspect, I decided to fold a fix for it into this PR as well.

#### Cookie update edge case

As the cookie is used to determine the default keyboard to use _globally_ on page reload, it does not make sense to update the cookie when changing the keyboard for controls using the "independent keyboard" mode.

#### Context-change during async keyboard activation

Setup:

- The web page has three controls:  A, B, and C.
- A and B both use the standard 'global' KMW keyboard setting, while C is configured with "independent keyboard" mode - its preferred keyboard is managed separately.
    - Active keyboard for "A" and "B":  "first"
    - Active keyboard for "C":  "special"

Sequence:
1. Current context:  control A.
2. The user requests to use a previously-unloaded keyboard, which is an asynchronous operation.
    - Requested keyboard: "second".
3. Experiencing some network delay, the user then swaps context to control C.
4. The network request completes, with the keyboard now fully loaded.

In this case, it is an error to set the keyboard active, as C does not follow the global keyboard setting.  It _is_ correct to set the 'global' setting to the new keyboard, though.  `setActiveKeyboard` would previously drop the keyboard change entirely, however.  
- Expected:  "A" and "B" should have "second" active (when revisited), with "C" still set to "special".
- Actual:  "A" and "B" are still set to "first", with "C" now set to "special".

Note that the same issue arises if we replace "C" with "B" and "A" with "C" in the "sequence" section above:
- Expected: "A" and "B" should still be set to "first", with "C" set to "second" (when revisited, as "C" was the active context when the change request was made)
- Actual: "A" and "B" would still be set to "first", with "C" still set to "special".

Handling this specific set of edge cases properly is the cause of much of the new additional handling code - `deferredKeyboardActivation` and `findAndPopActivation` in particular.

@keymanapp-test-bot skip

### CODE REVIEW

File: web/src/engine/main/src/contextManagerBase.ts [hunk 1]
```diff
@@ -1,11 +1,44 @@
...
+
+  /**
+   * This event is raised whenever a keyboard change is requested.  Calling the second parameter -
+   * `abortChange` - will abort the process and prevent the change.
+   *
+   * Note that if the keyboard has not been previously loaded, this event will be raised twice.
+   * 1. Before the keyboard is loaded into Keyman Engine for Web.
+   *     - Aborting at this point will prevent the keyboard from being loaded, with no network
+   *       request for the keyboard resource being triggered.
+   * 2. Once the keyboard is loaded, but before it is activated.
```
Comment (jahorton) [inline web/src/engine/main/src/contextManagerBase.ts:18]: Cross-reference with https://help.keyman.com/DEVELOPER/ENGINE/WEB/15.0/reference/events/kmw.beforekeyboardchange.

`abortChange` is the corresponding fix for #8482.

The "event raised twice" bit is based on the existing implementation of the event within KMW 16.0 and the timings thereof:

Check 1:
https://github.com/keymanapp/keyman/blob/f33f31521ae743ba6f9f0775965c86d1d28bfb79/web/src/engine/main/keyboards/kmwkeyboards.ts#L465
https://github.com/keymanapp/keyman/blob/f33f31521ae743ba6f9f0775965c86d1d28bfb79/web/src/engine/main/keyboards/kmwkeyboards.ts#L475-L476

(Where `_SetActiveKeyboard` is what triggers the keyboard-activation process)

Check 2:

https://github.com/keymanapp/keyman/blob/f33f31521ae743ba6f9f0775965c86d1d28bfb79/web/src/engine/main/keyboards/kmwkeyboards.ts#L713
https://github.com/keymanapp/keyman/blob/f33f31521ae743ba6f9f0775965c86d1d28bfb79/web/src/engine/main/keyboards/kmwkeyboards.ts#L752-L753
https://github.com/keymanapp/keyman/blob/f33f31521ae743ba6f9f0775965c86d1d28bfb79/web/src/engine/main/keyboards/kmwkeyboards.ts#L766-L768

(Upon the asynchronous loading of a newly-activating keyboard; `installKeyboard` is called by `_SetActiveKeyboard` for this case, listed as case 4 in the PR description)

**Comment by mcdurdin** [comment]
> **Context-change during async keyboard activation**
> ...
> `setActiveKeyboard` would previously drop the keyboard change entirely, however.

I see this as a valid and actually probably the most sensible approach, particularly because this is a real edge case scenario.

I think it's reasonable for the user to see their keyboard selection 'cancelled' if they didn't wait long enough for it to activate; if this allows you to simplify this code, I would encourage you do to that.

Furthermore, I note that a second attempt to select the keyboard later would be near instantaneous as the keyboard would still have been loaded after the first attempt, just not activated.

File: common/web/keyboard-processor/src/keyboards/loaders/nodeKeyboardLoader.ts [hunk 2]
```diff
@@ -23,15 +23,21 @@ export class NodeKeyboardLoader extends KeyboardLoaderBase {
...
-      // `fs` does not like 'file:///'; it IS "File System" oriented, after all, and wants a path, not a URI.
-      if(uri.indexOf('file:///') == 0) {
-        uri = uri.substring('file:///'.length);
-      }
-      const script = new vm.Script(fs.readFileSync(uri).toString());
       script.runInContext(this.harness._jsGlobal);
     } catch (err) {
-      return Promise.reject(err);
+      return Promise.reject(new Error(KeyboardLoaderBase.scriptErrorMessage(uri)));
     }
```
Comment (mcdurdin) [inline common/web/keyboard-processor/src/keyboards/loaders/nodeKeyboardLoader.ts:41]: It'd be good if the actual error message thrown (`err`) was not lost in these two exception handlers.

File: common/web/keyboard-processor/src/keyboards/keyboardLoaderBase.ts [hunk 3]
```diff
@@ -19,5 +20,48 @@ export default abstract class KeyboardLoaderBase {
...
+    let promise = this.loadKeyboardInternal(stub.filename, stub.id);
+
+    promise = promise.catch((err: Error) => {
+      if(err == KeyboardLoaderBase.scriptErrorMessage(stub.filename)) {
+        // Enhance the error message.
+        err.message = KeyboardLoaderBase.scriptErrorMessage(stub);
+      } else if(err == KeyboardLoaderBase.missingErrorMessage(stub.filename)) {
+        // Same thing here.
+        err.message = KeyboardLoaderBase.scriptErrorMessage(stub);
+      }
```
Comment (mcdurdin) [inline common/web/keyboard-processor/src/keyboards/keyboardLoaderBase.ts:34]: This is all a little magical. I don't really understand what you are doing here.

Also are you meaning line 33 to have `scriptErrorMessage` or `missingErrorMessage`?

File: common/web/keyboard-processor/src/keyboards/keyboardLoaderBase.ts [hunk 4]
```diff
@@ -19,5 +20,48 @@ export default abstract class KeyboardLoaderBase {
...
+  protected static missingErrorMessage(metadata: KeyboardProperties & { filename: string });
+  protected static missingErrorMessage(arg: string | (KeyboardProperties & { filename: string })) {
+    if(typeof arg == "string") {
+      const uri = arg;
+      return `Cannot find the keyboard at ${uri}.`
+    } else {
+      const stub = arg;
+      return `Cannot find the ${stub.name} keyboard for ${stub.langName} at ${stub.filename}.`;
+    }
+  }
```
Comment (mcdurdin) [inline common/web/keyboard-processor/src/keyboards/keyboardLoaderBase.ts:66]: These feel like a lot of syntactic sugar and typechecking for could essentially be separate error messages:

```
  protected static scriptUriErrorMessage(uri: string) {
    return `Error registering the keyboard script at ${uri}; it may contain an error.`;
  }

  protected static scriptStubErrorMessage(stub: KeyboardProperties & { filename: string }) {
    return `Error registering the ${stub.name} keyboard for ${stub.langName}; keyboard script at ${stub.filename} may contain an error.`;
  }
```

(That second one also feels painful because I don't really understand where filename is coming from in this instance)

File: web/src/app/browser/src/contextManager.ts [hunk 5]
```diff
@@ -58,4 +73,61 @@ export default class ContextManager extends ContextManagerBase {
...
     return false;
   }
+
+  /**
+   * Reflects the active 'target' upon which any `set activeKeyboard` operation will take place.
+   * When `null`, such operations will affect the global default; otherwise, such operations
+   * affect only the specified `target`.
+   */
+  protected get keyboardTarget(): OutputTarget {
+    // Remove `&& false` once the inlined section below is implemented.
```
Comment (mcdurdin) [inline web/src/app/browser/src/contextManager.ts:83]:
```suggestion
    // TODO: Remove `&& false` once the inlined section below is implemented.
```

Just makes it clearer that this is not yet finished

File: web/src/app/webview/src/contextManager.ts [hunk 6]
```diff
@@ -51,13 +51,50 @@ export default class ContextManager extends ContextManagerBase {
...
+    keyboardId: string,
+    languageCode?: string
+  ): {keyboard: Promise<Keyboard>, metadata: KeyboardStub} {
+    const originalKeyboard = this.activeKeyboard;
+    const activatingKeyboard = super.prepareKeyboardForActivation(keyboardId, languageCode);
+
+    // Probably isn't necessary at this point - osk.refreshLayout() exists - but
+    // it's best to keep it around for now and verify later.
+    if(originalKeyboard.metadata.id == activatingKeyboard.metadata.id) {
+      activatingKeyboard.keyboard = activatingKeyboard.keyboard.then((kbd) => {
```
Comment (mcdurdin) [inline web/src/app/webview/src/contextManager.ts:92]: Is this correct? `activatingKeyboard.keyboard` gets overwritten with a promise?

File: web/src/app/webview/src/contextManager.ts [hunk 7]
```diff
@@ -51,13 +51,50 @@ export default class ContextManager extends ContextManagerBase {
...
+
+  protected prepareKeyboardForActivation(
+    keyboardId: string,
+    languageCode?: string
+  ): {keyboard: Promise<Keyboard>, metadata: KeyboardStub} {
+    const originalKeyboard = this.activeKeyboard;
+    const activatingKeyboard = super.prepareKeyboardForActivation(keyboardId, languageCode);
+
+    // Probably isn't necessary at this point - osk.refreshLayout() exists - but
+    // it's best to keep it around for now and verify later.
```
Comment (mcdurdin) [inline web/src/app/webview/src/contextManager.ts:90]: I don't really understand this comment

**Review by mcdurdin — APPROVED** [review]
LGTM I guess? I am concerned about the additional complexity and file size incurred as evidenced by 250 new lines of code, and the fact that this is basically a complete rewrite of this functionality during what is supposed to be "just" a conversion from namespaces to modules. The setActiveKeyboard code _is_ fragile (and critical), and needs redesign, but I really think this kind of rewrite is better done in an isolated feature PR which can be thoroughly tested against the working baseline. As it stands, it is going to be very difficult to isolate changes in behavior in this PR chain when we finally get to user testing.

File: common/web/keyboard-processor/src/keyboards/loaders/nodeKeyboardLoader.ts (hunk 2 above)
Comment (jahorton) [reply common/web/keyboard-processor/src/keyboards/loaders/nodeKeyboardLoader.ts:41]: That's very fair.  I was looking for a 'simple' way to do that and came up short in my first pass, but I could always take an extra look.

File: common/web/keyboard-processor/src/keyboards/keyboardLoaderBase.ts (hunk 3 above)
Comment (jahorton) [reply common/web/keyboard-processor/src/keyboards/keyboardLoaderBase.ts:34]: Sounds like I should probably retool things a little bit to make the error-handling easier to follow, then.  I did feel that way when I did these changes, and figured I could triage to later... but if it's too difficult to follow in the PR that makes the change, then it probably should be made clearer here, now.

If you want details - the method that actually does the loading operates off of just the filename, without knowledge of the keyboard name and language name associated with the load request.  As a result, that method can only give the `stub.filename`-based version of the error.  _This_ method - the one loading from a stub - uses that, and so maps the filename-only error to the variant with richer feedback data.

That said, it may be best to retool things a bit so that there's no need to 'error map' in the first place.

**Comment by jahorton** [comment]
> LGTM I guess? I am concerned about the additional complexity and file size incurred as evidenced by 250 new lines of code, and the fact that this is basically a complete rewrite of this functionality during what is supposed to be "just" a conversion from namespaces to modules. The setActiveKeyboard code _is_ fragile (and critical), and needs redesign, but I really think this kind of rewrite is better done in an isolated feature PR which can be thoroughly tested against the working baseline. As it stands, it is going to be very difficult to isolate changes in behavior in this PR chain when we finally get to user testing.

I won't be able to get to it quite yet, but I could see about a more "rough and dirty" version using a module to establish semi-globals that would allow something closer to the original, then put this as its own, separate PR.  Do note this bit from the description up top:

>  These previously relied on explicit references to many of the higher-level, semi-global engine components and obtained their root object via the namespaced com.keyman.singleton global, so some cleanup here was quite necessary.

Since app/webview won't even _have_ equivalents to a few of the higher-level components needed for app/browser, the "rough and dirty" approach may also result in some "fun" typecasting down the line, some "fun" falsy-guards, and calls on names that will lack type specification due to being defined by `engine/main` consumers.  I'd wanted to avoid that mess, hence the approach I took initially.

**Comment by mcdurdin** [comment]
> Since app/webview won't even _have_ equivalents to a few of the higher-level components needed for app/browser, the "rough and dirty" approach may also result in some "fun" typecasting down the line, some "fun" falsy-guards, and calls on names that will lack type specification due to being defined by `engine/main` consumers. I'd wanted to avoid that mess, hence the approach I took initially.

Yep, it's always a difficult choice. We are here now, so let's keep this work, but try to minimize this level of refactoring as far as possible for further ES Module stuff :grin:

File: web/src/app/webview/src/contextManager.ts (hunk 6 above)
Comment (jahorton) [reply web/src/app/webview/src/contextManager.ts:92]: Note:  activat**ing**.  The keyboard is not yet activated - and quite possibly because it has not yet been loaded.  Thus, that field's typing is for `Promise<Keyboard>`, not `Keyboard`.

https://github.com/keymanapp/keyman/blob/3c185b31826864364d69999de708ebd993a853f3/web/src/engine/main/src/contextManagerBase.ts#L278-L281

So... "yes", except that we're less "overwriting" the `Promise` for the loading keyboard and more "extending" / "filtering" the `Promise`.

File: web/src/engine/main/src/contextManagerBase.ts (hunk 1 above)
Comment (jahorton) [reply web/src/engine/main/src/contextManagerBase.ts:18]: ... now that I've looked deeper into how the older style of KMW events operates... I think I was mistaken here.  That same verbiage is used for other events that clearly don't cancel any ongoing process whatsoever.  So... accidental new feature, if we leave it in.

I spent so much time to make that work, too...
